lib/UART/src/UART.cpp

```diff
@@ -17,10 +17,35 @@
         UCSR0A &= ~(1 << U2X0);
     #endif
 
-    UCSR0B = (1 << RXEN0) | (1 << TXEN0);   // Enable receiver and transmitter
+    cli();
+    UCSR0B = (1 << RXEN0) | (1 << TXEN0) | (1 << RXCIE0);   // Enable receiver, transmitter and receive interrupt
     UCSR0C = (1 << UCSZ01) | (1 << UCSZ00); // Set frame format: 8 data bits, 1 stop bit, no parity
+    sei();
+}
+
+#define RX_BUFFER_SIZE 64
+char rx_buffer[RX_BUFFER_SIZE];
+volatile uint8_t rx_index = 0;
+ISR(USART_RX_vect)
+{
+    char receivedChar = UDR0;
+    if (rx_index < RX_BUFFER_SIZE - 1)
+    {
+        rx_buffer[rx_index++] = receivedChar;
+        if (receivedChar == '\n' || receivedChar == '\r')
+        {
+            rx_buffer[rx_index] = '\0';
+            rx_index = 0;
+        }
+    }
+    else
+    {
+        rx_index = 0;
+    }
 }
```

Comment:
This implementation should be part of the `UART_receive_char`/`UART_receive_line` function instead!

```suggestion

uint8_queue_t rx_queue;

ISR(USART_RX_vect)
{
    char c = UDR0;
    add_to_queue(&rx_queue, (uint8_t) c);
}

#define RX_BUFFER_SIZE 64
static char line_buffer[RX_BUFFER_SIZE + 1];
static uint8_t line_index = 0;

void service_uart(void)
{
    if (queue_is_empty(&rx_queue)) {
        return;
    }

    v = dequeue(&rx_q);
    char c = v.value;

    bool is_eol = c == '\n' || c == '\r';
    if (is_eol) {
        line_buffer[line_index] = '\0';
        line_index = 0;
        return;
    }
    
    if (line_index >= RX_BUFFER_SIZE) {
        // ERROR: Buffer overflow
        return;
    }

    line_buffer[line_index] = c;
    line_index += 1;
}
```

Reply:
Whenever we get a full line, we should do something with it, like parse it.

Reply:
That is great. I am trying to implement it.
Question: 
service_uart() reads the char in the queue and moves it into the rx_buffer, which will eventually become the string we want to parse. To complete that string, service_uart() needs to be executed on every loop (poll), for every char in the queue until it reaches a \n or \r.
Could we instead do a while loop for every char in the service_uart()? or would that block the MCU from doing other tasks?

Reply:
The receive functionality was implemented as you stated. Now, when testing it, (by adding a 'UART_print_string()' to the service_uart() function), the functionality is working but you get that transmit delay when there is a count down (video that I shared with you on Slack).